File: F/2200-find-all-k-distant-indices-in-an-array/main.go
package main

func findKDistantIndices(nums []int, key int, k int) []int {

	ret := []int{}
	buf := make([]int, len(nums))
	idx := 0
	for idx < len(nums) {
		if nums[idx] != key {
			idx++
			continue
		}
		l, r := idx-k, idx+k
		if l < 0 {
			l = 0
		}
		if r >= len(nums) {
			r = len(nums) - 1
		}
		for l <= r {
			buf[l] = 1
			l++
		}
		idx++
	}
	for i := 0; i < len(nums); i++ {
		if buf[i] == 1 {
			ret = append(ret, i)
		}
	}

	return ret
}

File: F/2200-find-all-k-distant-indices-in-an-array/main.rs
impl Solution {
  pub fn find_k_distant_indices(nums: Vec<i32>, key: i32, k: i32) -> Vec<i32> {
    let mut start: i32 = 0;

    let mut ans: Vec<i32> = vec![];
    for (idx, v) in nums.iter().enumerate() {
      if *v == key {
        let idx = idx as i32;
        let end = (idx + k).min(nums.len() as i32 - 1);
        ((idx - k).max(start)..=end).for_each(|v| {
          ans.push(v);
        });
        start = end + 1;
      }
    }
    ans
  }
}
